presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoin.java

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - Join (left, right or full)
+ *       - left source
+ *       - right source
+ * </pre>
+ * Into:
+ * <pre>
+ * - TopN
+ *    - Join
+ *       - TopN (present if Join is left or outer, not already limited, and orderBy symbols come from left source)
+ *          - left source
+ *       - TopN (present if Join is right or outer, not already limited, and orderBy symbols come from right source)
+ *          - right source
+ * </pre>
+ */
+public class PushTopNThroughOuterJoin
+        implements Rule<TopNNode>
+{
+    private static final Capture<JoinNode> JOIN_CHILD = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN =
+            topN().with(source().matching(
+                    join().capturedAs(JOIN_CHILD).with(type().matching(type -> isLeftOrFullOuter(type) || isRightOrFullOuter(type)))));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(JOIN_CHILD);
+        TopNNode pushdownTopN = new TopNNode(context.getIdAllocator().getNextId(), joinNode, parent.getCount(), parent.getOrderingScheme(), PARTIAL);
+
+        List<Symbol> orderBySymbols = pushdownTopN.getOrderingScheme().getOrderBy();
+        PlanNode left = joinNode.getLeft();
+        PlanNode right = joinNode.getRight();
+
+        if (isLeftOrFullOuter(joinNode.getType())
+                && ImmutableSet.copyOf(left.getOutputSymbols()).containsAll(orderBySymbols)
+                && !isLimited(left, context.getLookup(), pushdownTopN.getCount())) {
+            return Result.ofPlanNode(
+                    parent.replaceChildren(ImmutableList.of(
+                            joinNode.replaceChildren(ImmutableList.of(
+                                    pushdownTopN.replaceChildren(ImmutableList.of(left)),
+                                    right)))));
+        }
+
+        if (isRightOrFullOuter(joinNode.getType())
+                && ImmutableSet.copyOf(right.getOutputSymbols()).containsAll(orderBySymbols)
+                && !isLimited(right, context.getLookup(), pushdownTopN.getCount())) {
+            return Result.ofPlanNode(
+                    parent.replaceChildren(ImmutableList.of(
+                            joinNode.replaceChildren(ImmutableList.of(
+                                    left,
+                                    pushdownTopN.replaceChildren(ImmutableList.of(right)))))));
+        }
+
+        return Result.empty();
+    }
+
+    private static boolean isLimited(PlanNode node, Lookup lookup, long limit)
+    {
+        Range<Long> cardinality = extractCardinality(node, lookup);
+        return cardinality.hasUpperBound() && cardinality.upperEndpoint() <= limit;
+    }
+
+    private static boolean isLeftOrFullOuter(JoinNode.Type type)
```

Comment:
I don't think these methods make the code more readable. They say exactly what the condition expresses, so they are not hiding complexity. I'd just inline them